### ISSUE TITLE
Serial number at 0 can not be decoded with #167

### ISSUE THREAD

**Issue by dinosaure** [mirleft/ocaml-x509#168]
During my tests, I decided to generate certificates in the fly via this code:
```ocaml
module Ca = struct
  open Rresult

  let prefix =
    X509.Distinguished_name.
      [ Relative_distinguished_name.singleton (CN "Fuzzer") ]

  let cacert_dn =
    X509.Distinguished_name.(
      prefix
      @ [ Relative_distinguished_name.singleton (CN "Ephemeral CA for fuzzer") ])

  let cacert_lifetime = Ptime.Span.v (365, 0L)
  let cacert_serial_number = Z.zero

  let make domain_name seed =
    Domain_name.of_string domain_name >>= Domain_name.host
    >>= fun domain_name ->
    let private_key =
      let seed = Cstruct.of_string (Base64.decode_exn ~pad:false seed) in
      let g = Mirage_crypto_rng.(create ~seed (module Fortuna)) in
      Mirage_crypto_pk.Rsa.generate ~g ~bits:2048 ()
    in
    let valid_from = Ptime.v (Ptime_clock.now_d_ps ()) in
    Ptime.add_span valid_from cacert_lifetime
    |> Option.to_result ~none:(R.msgf "End time out of range")
    >>= fun valid_until ->
    X509.Signing_request.create cacert_dn (`RSA private_key) >>= fun ca_csr ->
    let extensions =
      let open X509.Extension in
      let key_id =
        X509.Public_key.id X509.Signing_request.((info ca_csr).public_key)
      in
      let authority_key_id =
        ( Some key_id,
          X509.General_name.(singleton Directory [ cacert_dn ]),
          Some cacert_serial_number )
      in
      empty
      |> add Subject_alt_name
           ( true,
             X509.General_name.(
               singleton DNS [ Domain_name.to_string domain_name ]) )
      |> add Basic_constraints (true, (false, None))
      |> add Key_usage
           (true, [ `Digital_signature; `Content_commitment; `Key_encipherment ])
      |> add Subject_key_id (false, key_id)
      |> add Authority_key_id (false, authority_key_id)
    in
    X509.Signing_request.sign ~valid_from ~valid_until ~extensions
      ~serial:cacert_serial_number ca_csr (`RSA private_key) cacert_dn
    |> R.reword_error (R.msgf "%a" X509.Validation.pp_signature_error)
    >>= fun certificate ->
    let fingerprint = X509.Certificate.fingerprint `SHA256 certificate in
    let time () = Some (Ptime_clock.now ()) in
    let authenticator =
      X509.Authenticator.server_cert_fingerprint ~time ~hash:`SHA256
        ~fingerprint
    in
    Ok (certificate, `RSA private_key, authenticator)
end

let seed = Base64.encode_exn "foo"

let () =
  Mirage_crypto_rng_unix.initialize (module Mirage_crypto_rng.Fortuna);
  let cert, _pk, _authenticator = Result.get_ok (Ca.make "foo.bar" seed) in
  match X509.Certificate.(decode_der (encode_der cert)) with
  | Ok _ -> ()
  | Error (`Msg msg) -> failwith msg
```

You can compile this code with:
```shell
$ ocamlfind opt -linkpkg -package digestif.c,mirage-crypto-rng.unix,x509,rresult,ptime.clock.os main.ml
$ ./a.out
$ echo $?
0
```

Since I tried to move to the repository which contains all required packages where Cstruct was remove, I noticed an error about `X509.Certificate.decode_der`. If we switch to [no-cstruct](https://github.com/robur-coop/no-cstruct) and install `x509.1.0.0~beta`, we must update the code with this patch:
```diff
--- a.ml	2024-05-03 18:20:39.793393316 +0200
+++ b.ml	2024-05-03 18:21:03.681581905 +0200
@@ -17,7 +17,7 @@
     Domain_name.of_string domain_name >>= Domain_name.host
     >>= fun domain_name ->
     let private_key =
-      let seed = Cstruct.of_string (Base64.decode_exn ~pad:false seed) in
+      let seed = Base64.decode_exn ~pad:false seed in
       let g = Mirage_crypto_rng.(create ~seed (module Fortuna)) in
       Mirage_crypto_pk.Rsa.generate ~g ~bits:2048 ()
     in
```

And if we recompile the project, we got an error:
```shell
$ opam remote add no-cstruct https://github.com/robur-coop/no-cstruct.git
$ opam update
$ opam install x509.1.0.0~beta randomconv.0.2.0 mirage-crypto.1.0.0~beta mirage-crypto-rng.1.0.0~beta mirage-crypto-ec.1.0.0~beta mirage-crypto-pk.1.0.0~beta eqaf.1.0.0~beta pbkdf.2.0.0~beta hkdf.2.0.0~beta asn1-combinators.0.3.0
$ camlfind opt -linkpkg -package digestif.c,mirage-crypto-rng.unix,x509,rresult,ptime.clock.os main.ml
$ ./a.out
Fatal error: exception Failure("INTEGER: length 0")
```

If we set the `cacert_serial_number` to `Z.one`, everything works.

**Comment by hannesm** [comment]
Thanks for your report. So, the issue is manifold:
- (a) asn1-combinators removed the zarith and cstruct dependencies -- so a Asn.S.integer no longer is a Z.t (and Asn.S.integer used to care about encoding of negative values (see 93d267401b377a9b7f814eae397c37d35bd500f5 and bcb9f135c5f929d47082152ce5119da7c9182388) --> here X509 pretends a serial being an Z.t and an Asn.S.integer, but the encoding / decoding does not conform to ASN.1 rules (but uses Mirage_crypto_pk.Z_extra.to/of_octets_be directly)
- (b) should a serial number be a Z.t, really? in my experience, there's no arithmetic happening on it -- so it's fine to use a string directly
- (c) as seen in the EC signature woes, we should provide the necessary code to convert from "some string presenting an unsigned integer" to a "valid ASN.1 integer" (i.e. remove leading zeroes, potentially add a leading zero, potentially put a zero there, since zero-length integers are invalid as observed in this issue)
- (d) as well the code to convert some ASN.1 integer, treat it as unsigned integer in a sequence of octets should be somewhere in a blessed setting

The action plan may be:
- provide (c) and (d) directly in asn.1 combinators,
- improve the code here to use a string for serial number, but ensure we decode and encode good values.

Hope this answers your issue. I'm not sure about the timeline of all the "move-to-string/bytes" changes.

Also note from https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.2

> Note: Non-conforming CAs may issue certificates with serial numbers that are negative or zero.  Certificate users SHOULD be prepared to gracefully handle such certificates.

TL;DR: I guess it is fine to error on serial number being 0 or negative. But the error you observe is not nice, and we should fix this. :)

**Comment by hannesm** [comment]
After researching a bit more on the internet: we should allow decoding of certificates with a negative serial number, but should refuse to construct (encode) certificates with a negative number as serial. We should as well limit the serial number to 20 octets, as recommended in the RFC (again -- allow longer serial numbers being decoded, but avoid encoding such long serial number).

In the NSS ca certs file (part of ca-certs-nss package), there are several X.509 certificates with a serial of 0.

**Comment by hannesm** [comment]
with 67562f7 the serial is now an unsigned integer with maximum 20 octets. 0 is a valid value.
